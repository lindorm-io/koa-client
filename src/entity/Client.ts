import { ClientEvent } from "../enum";
import { EntityBase, EntityCreationError, IEntity, IEntityBaseOptions, TObject } from "@lindorm-io/core";

export interface IClient extends IEntity {
  approved: boolean;
  description: string;
  extra: TObject<any>;
  name: string;
  secret: string;
}

export interface IClientOptions extends IEntityBaseOptions {
  approved?: boolean;
  description?: string;
  extra?: TObject<any>;
  name?: string;
  secret?: string;
}

export class Client extends EntityBase implements IClient {
  private _approved: boolean;
  private _description: string;
  private _extra: TObject<any>;
  private _name: string;
  private _secret: string;

  constructor(options?: IClientOptions) {
    super(options);
    this._approved = options?.approved || false;
    this._description = options?.description || null;
    this._extra = options?.extra || {};
    this._name = options?.name || null;
    this._secret = options?.secret || null;
  }

  public get approved(): boolean {
    return this._approved;
  }
  public set approved(approved: boolean) {
    this._approved = approved;
    this.addEvent(ClientEvent.APPROVED_CHANGED, { approved: this._approved });
  }

  public get description(): string {
    return this._description;
  }
  public set description(description: string) {
    this._description = description;
    this.addEvent(ClientEvent.DESCRIPTION_CHANGED, { description: this._description });
  }

  public get extra(): TObject<any> {
    return this._extra;
  }
  public set extra(object: TObject<any>) {
    this._extra = object;
    this.addEvent(ClientEvent.EXTRA_CHANGED, { extra: this._extra });
  }

  public get name(): string {
    return this._name;
  }
  public set name(name: string) {
    this._name = name;
    this.addEvent(ClientEvent.NAME_CHANGED, { name: this._name });
  }

  public get secret(): string {
    return this._secret;
  }
  public set secret(secret: string) {
    this._secret = secret;
    this.addEvent(ClientEvent.SECRET_CHANGED, { secret: this._secret });
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== ClientEvent.CREATED) continue;
      throw new EntityCreationError("Client");
    }
    this.addEvent(ClientEvent.CREATED, {
      approved: this._approved,
      description: this._description,
      extra: this._extra,
      name: this._name,
      secret: this._secret,
      created: this._created,
      updated: this._updated,
    });
  }
}
