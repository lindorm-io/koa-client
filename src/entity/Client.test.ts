import MockDate from "mockdate";
import { Client } from "./Client";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));

MockDate.set("2020-01-01 08:00:00.000");

describe("Client.ts", () => {
  let client: Client;

  beforeEach(() => {
    client = new Client({
      approved: true,
      description: "description",
      extra: { emailAuthorizationUri: "https://lindorm.io" },
      name: "name",
      secret: "secret",
    });
  });

  test("should have all data", () => {
    expect(client).toMatchSnapshot();
  });

  test("should have optional data", () => {
    client = new Client();

    expect(client).toMatchSnapshot();
  });

  test("should create", () => {
    client.create();
    expect(client.events).toMatchSnapshot();
  });

  test("should get/set approved", () => {
    expect(client.approved).toBe(true);

    client.approved = false;

    expect(client.approved).toBe(false);
    expect(client.events).toMatchSnapshot();
  });

  test("should get/set extra", () => {
    expect(client.extra).toStrictEqual({ emailAuthorizationUri: "https://lindorm.io" });

    client.extra = { emailAuthorizationUri: "https://lindorm.io/new" };

    expect(client.extra).toStrictEqual({ emailAuthorizationUri: "https://lindorm.io/new" });
    expect(client.events).toMatchSnapshot();
  });

  test("should get/set description", () => {
    expect(client.description).toBe("description");

    client.description = "new-description";

    expect(client.description).toBe("new-description");
    expect(client.events).toMatchSnapshot();
  });

  test("should get/set name", () => {
    expect(client.name).toBe("name");

    client.name = "new-name";

    expect(client.name).toBe("new-name");
    expect(client.events).toMatchSnapshot();
  });

  test("should get/set secret", () => {
    expect(client.secret).toBe("secret");

    client.secret = "new-secret";

    expect(client.secret).toBe("new-secret");
    expect(client.events).toMatchSnapshot();
  });
});
