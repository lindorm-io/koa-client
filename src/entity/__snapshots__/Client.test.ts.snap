// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client.ts should create 1`] = `
Array [
  Object {
    "date": 2020-01-01T07:00:00.000Z,
    "name": "client_created",
    "payload": Object {
      "approved": true,
      "created": 2020-01-01T07:00:00.000Z,
      "description": "description",
      "extra": Object {
        "emailAuthorizationUri": "https://lindorm.io",
      },
      "name": "name",
      "secret": "secret",
      "updated": 2020-01-01T07:00:00.000Z,
    },
  },
]
`;

exports[`Client.ts should get/set approved 1`] = `
Array [
  Object {
    "date": 2020-01-01T07:00:00.000Z,
    "name": "client_approved_changed",
    "payload": Object {
      "approved": false,
    },
  },
]
`;

exports[`Client.ts should get/set description 1`] = `
Array [
  Object {
    "date": 2020-01-01T07:00:00.000Z,
    "name": "client_description_changed",
    "payload": Object {
      "description": "new-description",
    },
  },
]
`;

exports[`Client.ts should get/set extra 1`] = `
Array [
  Object {
    "date": 2020-01-01T07:00:00.000Z,
    "name": "client_extra_changed",
    "payload": Object {
      "extra": Object {
        "emailAuthorizationUri": "https://lindorm.io/new",
      },
    },
  },
]
`;

exports[`Client.ts should get/set name 1`] = `
Array [
  Object {
    "date": 2020-01-01T07:00:00.000Z,
    "name": "client_name_changed",
    "payload": Object {
      "name": "new-name",
    },
  },
]
`;

exports[`Client.ts should get/set secret 1`] = `
Array [
  Object {
    "date": 2020-01-01T07:00:00.000Z,
    "name": "client_secret_changed",
    "payload": Object {
      "secret": "new-secret",
    },
  },
]
`;

exports[`Client.ts should have all data 1`] = `
Client {
  "_approved": true,
  "_created": 2020-01-01T07:00:00.000Z,
  "_description": "description",
  "_events": Array [],
  "_extra": Object {
    "emailAuthorizationUri": "https://lindorm.io",
  },
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_name": "name",
  "_secret": "secret",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;

exports[`Client.ts should have optional data 1`] = `
Client {
  "_approved": false,
  "_created": 2020-01-01T07:00:00.000Z,
  "_description": null,
  "_events": Array [],
  "_extra": Object {},
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_name": null,
  "_secret": null,
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;
